qwerty_ast/src/typecheck.rs

```diff
@@ -160,17 +216,37 @@ pub fn typecheck_expr(expr: &Expr, env: &mut TypeEnv) -> Result<Type, TypeError>
 
         Expr::UnitLiteral { dbg: _ } => Ok(Type::UnitType),
 
-        Expr::Adjoint { func, dbg: _ } => {
+        Expr::Adjoint { func, dbg } => {
             // Adjoint should be a function type (unitary/quantum), not classical.
             let func_ty = typecheck_expr(func, env)?;
-            // TODO: Enforce Qwerty adjoint typing rules.
-            Ok(func_ty)
+
+            match func_ty {
+                Type::RevFuncType { in_out_ty } => Ok(Type::RevFuncType { in_out_ty }),
+                Type::FuncType { .. } => {
+                    // Classical functions cannot have an adjoint
+                    Err(TypeError {
+                        kind: TypeErrorKind::InvalidType(format!(
+                            "Cannot take adjoint of non-reversible function: {:?}",
```

Comment:
Thanks for efforts on these error messages. I think it will pay off when we get students to use this